File: mahaam-api-java/src/main/java/mahaam/feat/task/TaskController.java
package mahaam.feat.task;

import java.util.List;
import java.util.UUID;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.DELETE;
import jakarta.ws.rs.FormParam;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.PATCH;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.PathParam;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import mahaam.infra.Json;
import mahaam.infra.Rule;

public interface TaskController {
	Response create(UUID planId, String title);

	Response delete(UUID planId, UUID id);

	Response updateDone(UUID planId, UUID id, Boolean done);

	Response updateTitle(UUID id, String title);

	Response reOrder(UUID planId, int oldOrder, int newOrder);

	Response getMany(UUID planId);
}

@ApplicationScoped
@Path("/plans/{planId}/tasks")
@Consumes(MediaType.APPLICATION_FORM_URLENCODED)
@Produces(MediaType.APPLICATION_JSON)
class DefaultTaskController implements TaskController {

	@Inject
	TaskService taskService;

	@POST
	public Response create(@PathParam("planId") UUID planId, @FormParam("title") String title) {
		Rule.required(planId, "planId");
		Rule.required(title, "title");

		UUID id = taskService.create(planId, title);
		return Response.status(Response.Status.CREATED).entity(Json.toString(id)).build();
	}

	@DELETE
	@Path("/{id}")
	public Response delete(@PathParam("planId") UUID planId, @PathParam("id") UUID id) {
		Rule.required(planId, "planId");
		Rule.required(id, "id");

		taskService.delete(planId, id);
		return Response.noContent().build();
	}

	@PATCH
	@Path("/{id}/done")
	public Response updateDone(
			@PathParam("planId") UUID planId,
			@PathParam("id") UUID id,
			@FormParam("done") Boolean done) {
		Rule.required(planId, "planId");
		Rule.required(id, "id");
		Rule.required(done, "done");

		taskService.updateDone(planId, id, done);
		return Response.ok().build();
	}

	@PATCH
	@Path("/{id}/title")
	public Response updateTitle(@PathParam("id") UUID id, @FormParam("title") String title) {
		Rule.required(id, "id");
		Rule.required(title, "title");

		taskService.updateTitle(id, title);
		return Response.ok().build();
	}

	@PATCH
	@Path("/reorder")
	public Response reOrder(
			@PathParam("planId") UUID planId,
			@FormParam("oldOrder") int oldOrder,
			@FormParam("newOrder") int newOrder) {
		Rule.required(planId, "planId");
		Rule.required(oldOrder, "oldOrder");
		Rule.required(newOrder, "newOrder");

		taskService.reOrder(planId, oldOrder, newOrder);
		return Response.ok().build();
	}

	@GET
	public Response getMany(@PathParam("planId") UUID planId) {
		Rule.required(planId, "planId");

		List<Task> result = taskService.getList(planId);
		return Response.ok(Json.toString(result)).build();
	}
}

File: mahaam-api-cs/Src/Feat/Task/TaskController.cs

using System.Net;
using Mahaam.Infra;
using Microsoft.AspNetCore.Mvc;

namespace Mahaam.Feat.Tasks;

public interface ITaskController
{
	Task<IActionResult> Create(Guid planId, string title);
	Task<IActionResult> Delete(Guid planId, Guid id);
	Task<IActionResult> UpdateDone(Guid planId, Guid id, bool done);
	Task<IActionResult> UpdateTitle(Guid id, string title);
	Task<IActionResult> ReOrder(Guid planId, int oldOrder, int newOrder);
	Task<IActionResult> GetMany(Guid planId);
}

[ApiController]
[Route("plans/{planId}/tasks")]
public class TaskController(ITaskService taskService) : ControllerBase, ITaskController
{
	private readonly ITaskService _taskService = taskService;
	[HttpPost]
	public async Task<IActionResult> Create(Guid planId, [FromForm] string title)
	{
		Rule.Required(planId, "planId");
		Rule.Required(title, "title");
		var id = await _taskService.Create(planId, title);
		return Created($"/plans/{planId}/tasks/{id}", id);
	}

	[HttpDelete]
	[Route("{id}")]
	public async Task<IActionResult> Delete(Guid planId, Guid id)
	{
		Rule.Required(planId, "planId");
		Rule.Required(id, "id");
		await _taskService.Delete(planId, id);
		return NoContent();
	}

	[HttpPatch]
	[Route("{id}/done")]
	public async Task<IActionResult> UpdateDone(Guid planId, Guid id, [FromForm] bool done)
	{
		Rule.Required(planId, "planId");
		Rule.Required(id, "id");
		Rule.Required(done, "done");
		await _taskService.UpdateDone(planId, id, done);
		return Ok();
	}

	[HttpPatch]
	[Route("{id}/title")]
	public async Task<IActionResult> UpdateTitle(Guid id, [FromForm] string title)
	{
		Rule.Required(id, "id");
		Rule.Required(title, "title");
		await _taskService.UpdateTitle(id, title);
		return Ok();
	}

	[HttpPatch]
	[Route("reorder")]
	public async Task<IActionResult> ReOrder(Guid planId, [FromForm] int oldOrder, [FromForm] int newOrder)
	{
		Rule.Required(planId, "planId");
		Rule.Required(oldOrder, "oldOrder");
		Rule.Required(newOrder, "newOrder");
		await _taskService.ReOrder(planId, oldOrder, newOrder);
		return Ok();
	}

	[HttpGet]
	[Route("")]
	public async Task<IActionResult> GetMany(Guid planId)
	{
		Rule.Required(planId, "planId");
		var result = await _taskService.GetList(planId);
		return Ok(result);
	}

}
